torch/jit/_isinstance.py

```diff
@@ -0,0 +1,109 @@
+from typing import List, Dict, Tuple, Union, Optional
+
+
+def get_origin(the_type):
+    return getattr(the_type, "__origin__", None)
+
+
+def get_args(the_type):
+    return getattr(the_type, "__args__", None)
+
+
+def check_args_exist(the_type):
+    if the_type is List or the_type is list:
+        raise RuntimeError(
+            "Attempted to use List without a "
+            "contained type. Please add a contained type, e.g. "
+            "List[int]"
+        )
+    elif the_type is Tuple or the_type is tuple:
+        raise RuntimeError(
+            "Attempted to use Tuple without a "
+            "contained type. Please add a contained type, e.g. "
+            "Tuple[int]"
+        )
+    elif the_type is Dict or the_type is dict:
+        raise RuntimeError(
+            "Attempted to use Dict without "
+            "contained types. Please add contained type, e.g. "
+            "Dict[int, int]"
+        )
+    elif the_type is None or the_type is Optional:
+        raise RuntimeError(
+            "Attempted to use Optional without a "
+            "contained type. Please add a contained type, e.g. "
+            "Optional[int]"
+        )
+
+
+def generics_checker(the_obj, the_type):
+    origin_type = get_origin(the_type)
+    check_args_exist(the_type)
+    if origin_type is None:
+        pass
+    elif origin_type is list or origin_type is List:
+        if isinstance(the_obj, list):
+            for el in the_obj:
+                # check if nested generics, ex: List[List[str]]
+                arg_type = get_args(the_type)[0]
+                arg_origin = get_origin(arg_type)
+                if arg_origin:  # processes nested generics, ex: List[List[str]]
+                    if not generics_checker(el, arg_type):
+                        return False
+                elif not isinstance(el, arg_type):
+                    return False
+        else:
+            return False
+    elif origin_type is dict or origin_type is Dict:
+        if isinstance(the_obj, dict):
+            key_type = get_args(the_type)[0]
+            val_type = get_args(the_type)[1]
+            for key, val in the_obj.items():
+                # check if keys are of right type
+                if not isinstance(key, key_type):
+                    return False
+                val_origin = get_origin(val_type)
+                if val_origin:
+                    if not generics_checker(val, val_type):
+                        return False
+                elif not isinstance(val, val_type):
+                    return False
+        else:
+            return False
+    elif origin_type is Union:  # TODO actually handles Optional Case
```

Comment:
Calling `getattr(the_type, "__origin__", None)` on an Optional[str] type returns the type of typing.Union instead of typing.Option. Whoever wrote the is_optional logic in _jit_internal.py did some similarly odd logic to address this